src/fsharp/FSharp.Core/array.fs

```diff
@@ -408,32 +423,53 @@ namespace Microsoft.FSharp.Collections
             let rec loop i = i >= len1 || (f.Invoke(array1.[i], array2.[i]) && loop (i+1))
             loop 0
 
-        [<CompiledName("GroupBy")>]
-        let groupBy keyf (array: 'T[]) =
-            checkNonNull "array" array
-            let dict = new Dictionary<RuntimeHelpers.StructBox<'Key>,ResizeArray<'T>>(RuntimeHelpers.StructBox<'Key>.Comparer)
+        let inline groupByImpl (comparer:IEqualityComparer<'SafeKey>) (keyf:'T->'SafeKey) (getKey:'SafeKey->'Key) (array: 'T[]) =
+            let dict = Dictionary<_,ResizeArray<_>> comparer
+
+            // Previously this was 1, but I think this is rather stingy, considering that we are alreadying paying
+            // for at least a key, the ResizeArray reference, which includes an array reference, an Entry in the
+            // Dictionary, plus any empty space in the Dictionary of unfilled hash buckets. Having it larger means
+            // that we won't be having as many re-allocations. The ResizeArray is destroyed at the end anyway.
+            let initialBucketSize = 4
```

Comment:
`List<'t>` defaults to 0 capacity, then grows to size 4 on the first `Add`.  The initial length-0 backing store is a shared static guy, so no allocation is needed. Thus using explicit initial capacity of 4 results in exactly the same allocation profile as using the default constructor :smile:

Using 1 as the default capacity potentially pays off if most buckets wind up with only 1 element, but as you mention it's worse if it has to grow a couple times.  Without some serious usage pattern research I don't think there's much point in guessing about something like that.

So my recommendation would be to either leave it as-is, or use default ctor `ResizeArray()`